File: Leetcode Question/242 Valid Anagram.java
/*Leetcode 242. Valid Anagram

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.


Example 1:
Input: s = "anagram", t = "nagaram"
Output: true

Example 2:
Input: s = "rat", t = "car"
Output: false
 

Constraints:
1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.
*/

class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()){
            return false;
        }
        
        // HashMap <Key,Value> map;
        HashMap<Character, Integer> map = new HashMap<>();
        
        for(int i = 0; i < s.length(); i++){
            // if s have same character, hashmap value + 1
            map.put(s.charAt(i), map.getOrDefault(s.charAt(i), 0) + 1);  
            // if t have same character, hashmap value - 1
            map.put(t.charAt(i), map.getOrDefault(t.charAt(i), 0) - 1);
        } // so if s and t have the same elements and the same quantity, all values of the hashmap should be 0
        
        //The Java HashMap getOrDefault() method returns the specified default value if the mapping for the specified key is not found in the hashmap. 
     //Otherwise, the method returns the value corresponding to the specified key. 
        
        
        // HashMap <Key,Value> map; => So in here, we use value to check whether is 's' and 't' same elements and the same quantity.
        for(int value : map.values()){
            if(value != 0){ // if there is a element value is equal 0, it means not "Valid Anagram"
                return false; //not "Valid Anagram"
            }
        }
        return true;
    }
}

File: Interview OA/BeaconFire OA/242 Valid Anagram.js
/* Leetcode 242. Valid Anagram

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.


Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false
 

Constraints:

1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.
 

Follow up: What if the inputs contain Unicode characters? How would you adapt your solution to such a case?
*/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */

var isAnagram = function(s, t) {
    if(s.length !== t.length){
        return false;
    }

    let map = {}; //hashmap
    //used hashmap to save the characters from string t and s, and count the freqence time
    for(let i = 0; i < s.length; i++){
        map[s[i]] = map[s[i]] ? map[s[i]] + 1 : 1;
        map[t[i]] = map[t[i]] ? map[t[i]] - 1 : -1;
    }

    for(let key in map){
        if(map[key] !== 0){
            return false;
        }
    }

    return true;
};
